### PR TITLE
Added a flag to not send goal in BtActionNode

### PR DESCRIPTION
**Pull request by owen7900**
Signed-off-by: Owen Hooper <17ofh@queensu.ca>

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3150  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo |

---

## Description of contribution in a few bullet points

* I added a parameter to allow an implementing developer to prevent the BtActionNode from sending a goal
* Parameter gets reset before `on_tick` is called
* Parameter is checked before both calls to `send_new_goal`

## Description of documentation updates required from your changes
* Needs to get added to the migration guide
* Needs to be added to the table in writing_new_bt_plugin 

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp [hunk 1]
```diff
@@ -188,10 +188,16 @@ class BtActionNode : public BT::ActionNodeBase
...
       setStatus(BT::NodeStatus::RUNNING);
 
+      // reset the flag to send the goal or not, allowing the user the option to set it in on_tick
+      should_send_goal_ = true;
+
       // user defined callback
       on_tick();
 
-      send_new_goal();
+      if(should_send_goal_)
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:197]: Linting says this should be `if (should_send_goal_) {`. You can reproduce locally by running `ament_cpplint` and `ament_uncrustify`

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp [hunk 2]
```diff
@@ -223,7 +229,7 @@ class BtActionNode : public BT::ActionNodeBase
         feedback_.reset();
 
         auto goal_status = goal_handle_->get_status();
-        if (goal_updated_ && (goal_status == action_msgs::msg::GoalStatus::STATUS_EXECUTING ||
+        if (goal_updated_ && should_send_goal_ && (goal_status == action_msgs::msg::GoalStatus::STATUS_EXECUTING ||
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:232]: I think this is for preemption which we'd always want to run? The `on_wait_for_result()` function may choose or not to update `goal_updated_` so I think that would meet the need here. I wouldn't think someone would set `goal_updated_` to true while also setting `should_send_goal_` to false after ticking the node. That feels like they meet the same intention just in the initial tick vs preemption during execution contexts

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Thanks!

**Comment by SteveMacenski** [comment]
Any update?

**Comment by owen7900** [comment]
@SteveMacenski  Got busy. Made the changes.

**Review by SteveMacenski — APPROVED** [review]

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp [hunk 3]
```diff
@@ -188,10 +188,15 @@ class BtActionNode : public BT::ActionNodeBase
       // setting the status to RUNNING to notify the BT Loggers (if any)
       setStatus(BT::NodeStatus::RUNNING);
 
+      // reset the flag to send the goal or not, allowing the user the option to set it in on_tick
+      should_send_goal_ = true;
+
       // user defined callback
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:194]:
```suggestion
      // user defined callback, may modify "should_send_goal_".
```

**Comment by SteveMacenski** [comment]
Thanks!

**Comment by tonynajjar** [comment]
When I set `should_send_goal_` to false in my `on_tick()`, the navigator crashes, here is the log with backward_ros:

```
[amr_navigate_to_pose-6]          77:   template<typename _Res, typename _MemPtr, typename _Tp>
[amr_navigate_to_pose-6] #6    Source "/code/ros2_ws/install/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp", line 237, in executeCallback [0x7f841716e71f]
[amr_navigate_to_pose-6]         234:     };
[amr_navigate_to_pose-6]         235: 
[amr_navigate_to_pose-6]         236:   // Execute the BT that was previously created in the configure step
[amr_navigate_to_pose-6]       > 237:   nav2_behavior_tree::BtStatus rc = bt_->run(tree_, on_loop, is_canceling, bt_loop_duration_);
[amr_navigate_to_pose-6]         238: 
[amr_navigate_to_pose-6]         239:   // Make sure that the Bt is not in a running state from a previous execution
[amr_navigate_to_pose-6]         240:   // note: if all the ControlNodes are implemented correctly, this is not needed.
[amr_navigate_to_pose-6] #5    Object "/code/ros2_ws/build/nav2_behavior_tree/libnav2_behavior_tree.so", at 0x7f841652f552, in nav2_behavior_tree::BehaviorTreeEngine::run(BT::Tree*, std::function<void ()>, std::function<bool ()>, std::chrono::duration<long, std::ratio<1l, 1000l> >)
[amr_navigate_to_pose-6] #4    Object "/opt/ros/humble/lib/libbehaviortree_cpp_v3.so", at 0x7f8417561745, in BT::TreeNode::executeTick()
[amr_navigate_to_pose-6] #3    Object "/opt/ros/humble/lib/libbehaviortree_cpp_v3.so", at 0x7f8417575b3b, in BT::SequenceNode::tick()
[amr_navigate_to_pose-6] #2    Object "/opt/ros/humble/lib/libbehaviortree_cpp_v3.so", at 0x7f8417561745, in BT::TreeNode::executeTick()
[amr_navigate_to_pose-6] #1    Object "/code/ros2_ws/build/amr_behavior_tree/libset_lift_level_action_bt_node.so", at 0x7f8410188861, in nav2_behavior_tree::BtActionNode<control_msgs::action::SingleJointPosition>::tick()
[amr_navigate_to_pose-6] #0    Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f841697ff74, in pthread_mutex_lock
[amr_navigate_to_pose-6] Segmentation fault (Address not mapped to object [0xb0])
[ERROR] [amr_navigate_to_pose-6]: process has died [pid 2432571, exit code -11, cmd '/code/ros2_ws/install/amr_bt_navigator/lib/amr_bt_navigator/amr_navigate_to_pose --ros-args -r __node:=amr_navigate_to_pose -p use_sim_time:=True --params-file /tmp/unaliased_2knhja99'].
``` 

As you can probably see a lot of the code in there in custom so I'm not 100% sure that the issue is on Nav2 side but a test for this would have been really useful to confirm or deny this. Do you see any obvious reason for this crash?

**Pull request merged**